# テスト戦略ガイド（テストファースト開発）

**目的**: Phase A2で発生した統合テスト課題の根本的解決・品質向上・Claude主体開発の最適化  
**適用範囲**: 全Phase・全Step  
**作成日**: 2025-07-21  
**最終更新**: 2025-07-21  

## 🔴 テストファースト開発の必須化

### 基本方針
**目的**: Phase A2で発生した統合テスト課題の根本的解決・品質向上・Claude主体開発の最適化

### 🔴 必須実行ルール
```
各Step実装時に以下を必須実行:
□ テストケース事前設計・実装
□ 実装前のテスト失敗確認（Red）
□ 最小限実装によるテスト成功（Green）
□ リファクタリング・品質向上（Refactor）
□ テストカバレッジ80%以上維持
```

### 仕様ベースRed-Green-Refactorサイクル
1. **仕様確認**: 機能仕様書の該当箇所を精読し、テストすべき内容を特定
2. **Red**: 仕様に基づいて失敗するテストを書く（肯定的仕様・否定的仕様両方）
3. **Green**: 仕様を満たす最小限のコードを書く
4. **Refactor**: コードを改善し、仕様準拠とテスト成功を維持

## 📊 テスト層別戦略

### 🟢 単体テスト（各Step実装と同時）

**対象**: F# Domain・Application層、C# Infrastructure・Web層

**手法**: テストファースト（Red→Green→Refactor）

**品質基準**: 
- 各メソッド・関数のテストケース作成
- エッジケース・異常系テストの包括的実装
- F# FsUnit、C# xUnit活用

**仕様準拠実装パターン**:
```fsharp
// F# Domain層テスト例
// 仕様書3.2.1準拠: メールアドレス形式検証
[<Test>]
let ``Email.create should validate correct email format per spec 3.2.1`` () =
    // Arrange
    let validEmail = "test@example.com"
    
    // Act & Assert
    match Email.create validEmail with
    | Ok email -> email.Value |> should equal validEmail
    | Error _ -> failwith "Valid email should not fail"

// 仕様書3.2.1準拠: 無効なメールアドレス拒否
[<Test>]
let ``Email.create should reject invalid format per spec 3.2.1`` () =
    // Arrange
    let invalidEmail = "invalid-email"
    
    // Act & Assert
    match Email.create invalidEmail with
    | Ok _ -> failwith "Invalid email should fail validation"
    | Error error -> error |> should contain "Invalid email format"
```

**否定的仕様テストパターン**:
```fsharp
// 仕様書2.1.1準拠: ログイン失敗によるロックアウト機構は設けない
[<Test>]
let ``Login should not lock account after multiple failures per spec 2.1.1`` () =
    // Arrange - 10回の連続ログイン失敗を準備
    let loginAttempts = [1..10]
    let mutable finalResult = None
    
    // Act - 10回失敗後も正しいパスワードでログイン可能
    for _ in loginAttempts do
        AuthService.login "user@example.com" "wrongpassword" |> ignore
    
    finalResult <- Some (AuthService.login "user@example.com" "correctpassword")
    
    // Assert - ロックアウトされていないことを確認
    match finalResult with
    | Some (Ok _) -> () // 成功：仕様準拠
    | Some (Error msg) when msg.Contains("locked") -> 
        failwith "Account should not be locked per spec 2.1.1"
    | _ -> failwith "Unexpected result"
```

```csharp
// C# Infrastructure層テスト例
[Fact]
public async Task GetByIdAsync_ShouldReturnUser_WhenUserExists()
{
    // Arrange
    var userId = Guid.NewGuid();
    var expectedUser = CreateTestUser(userId);
    
    // Act
    var result = await _userRepository.GetByIdAsync(userId);
    
    // Assert
    Assert.NotNull(result);
    Assert.Equal(expectedUser.Id, result.Id);
}
```

### 🟡 統合テスト（レイヤー統合時）

**対象**: F#↔C#境界・DB接続・ASP.NET Core Identity連携

**手法**: WebApplicationFactory・In-Memory Database活用

**品質基準**:
- 層間データフロー完全検証
- Phase A2で発生したHTTP 500エラー類の事前発見・解決
- 認証・認可フロー完全検証

**実装パターン**:
```csharp
public class UserIntegrationTests : IClassFixture<WebApplicationFactory<Program>>
{
    private readonly WebApplicationFactory<Program> _factory;
    
    public UserIntegrationTests(WebApplicationFactory<Program> factory)
    {
        _factory = factory;
    }
    
    [Fact]
    public async Task CreateUser_ShouldReturnSuccess_WhenValidData()
    {
        // Arrange
        var client = _factory.CreateClient();
        var createUserDto = new CreateUserDto { /* ... */ };
        
        // Act
        var response = await client.PostAsJsonAsync("/api/users", createUserDto);
        
        // Assert
        response.EnsureSuccessStatusCode();
        var createdUser = await response.Content.ReadFromJsonAsync<UserDto>();
        Assert.NotNull(createdUser);
    }
}
```

### Phase A2課題の具体的解決
**課題**: WebApplicationFactory統合テストでHTTP 500エラー（6テスト失敗）

**解決策**:
1. **ASP.NET Core Identity In-Memory DB対応設定**
2. **TestInitialDataService完全実装**
3. **UserManager/RoleManager/AuthenticationStateProvider適切なモック**
4. **DbContext依存関係整理・重複解消**

### 🔴 E2Eテスト（UI実装完了時）

**対象**: Blazor Server UI・ユーザーシナリオ・ブラウザ動作

**手法**: Playwright・Selenium活用検討

**品質基準**:
- 主要ユーザーシナリオの自動化
- ブラウザ互換性確認
- レスポンシブデザイン検証

**実装パターン**:
```csharp
[Test]
public async Task UserLogin_ShouldNavigateToHome_WhenCredentialsValid()
{
    // Arrange
    await using var browser = await Playwright.Chromium.LaunchAsync();
    var page = await browser.NewPageAsync();
    
    // Act
    await page.GotoAsync("https://localhost:5001/login");
    await page.FillAsync("#email", "test@example.com");
    await page.FillAsync("#password", "TestPassword123!");
    await page.ClickAsync("#login-button");
    
    // Assert
    await page.WaitForURLAsync("**/home");
    var title = await page.TextContentAsync("h1");
    Assert.Contains("Welcome", title);
}
```

### 🟣 受け入れテスト（Phase完了前）

**対象**: ビジネス要件・ユーザビリティ・パフォーマンス

**手法**: 手動テスト・品質チェックリスト

**品質基準**:
- 要件仕様書との完全準拠確認
- ユーザビリティ基準クリア
- パフォーマンス基準満足

## 🔄 Step実行時テスト統合

### Step開始時（テストファースト準備）
1. **テスト計画策定**: Step目標に基づくテストケース設計
2. **テスト環境準備**: 必要なテストデータ・モック・設定準備
3. **テスト実装開始**: 機能実装前のテストケース作成

### Step実行中（Red-Green-Refactorサイクル）
1. **Red**: 失敗するテストの実装・確認
2. **Green**: 最小限実装によるテスト成功
3. **Refactor**: コード品質向上・テスト成功維持

### Step終了時テスト品質確認

**🔴 Step終了時必須確認**:
```
□ 新規実装機能の単体テスト完成・成功
□ 統合テスト実行・課題解決完了
□ テストカバレッジ80%以上維持確認
□ 全テスト成功状態でのStep完了
□ 次Stepテスト環境・前提条件確認
```

## 📋 技術スタック別実装ガイド

### F# Domain・Application層
- **テストフレームワーク**: NUnit + FsUnit
- **モッキング**: Moq for F#
- **重点**: 純粋関数・型安全性・ビジネスルール

### C# Infrastructure・Web層
- **テストフレームワーク**: xUnit
- **統合テスト**: WebApplicationFactory
- **モッキング**: Moq
- **重点**: DB操作・HTTP通信・認証認可

### Blazor Server UI
- **コンポーネントテスト**: bUnit
- **E2Eテスト**: Playwright.NET
- **重点**: ユーザーインタラクション・レスポンシブ対応

## 🎯 品質保証指標

### テストカバレッジ目標
- **単体テスト**: 80%以上
- **統合テスト**: 主要フロー100%
- **E2Eテスト**: 重要ユーザーシナリオ100%

### パフォーマンス基準
- **単体テスト実行時間**: 10秒以内（全テスト）
- **統合テスト実行時間**: 60秒以内（全テスト）
- **E2Eテスト実行時間**: 5分以内（全シナリオ）

### 品質ゲート
- **全テスト成功**: Step完了の必須条件
- **カバレッジ維持**: 80%以上維持必須
- **パフォーマンス劣化**: 前回比10%以内

## 🔧 実装支援ツール

### 推奨ツール
- **テストランナー**: dotnet test
- **カバレッジ**: coverlet.collector
- **レポート**: ReportGenerator
- **CI統合**: GitHub Actions

### 開発環境統合
```json
// .vscode/settings.json
{
    "dotnet.test.explorer.enabled": true,
    "dotnet.test.explorer.autoRun": true,
    "coverage-gutters.enableFileOnSave": true
}
```

## 🏢 TDD実践のための組織設計テンプレート

### フィーチャーチーム制（推奨）
実装とテストを分離せず、機能単位でチームを編成し、各チームがRed-Green-Refactorサイクルを完結できる体制。

```markdown
#### Feature Team 1: [機能名]
**責任範囲**: [機能の詳細説明]
**TDD実践計画**:
- Red Phase: [作成するテストケース一覧]
- Green Phase: [実装する最小機能一覧]
- Refactor Phase: [改善予定項目一覧]
**技術スタック**: [使用する技術・フレームワーク]
```

### TDDサイクル実行計画テンプレート
```markdown
## TDDサイクル実行計画
### 0:00-0:20 [Red Phase - テスト作成]
├── 全チーム: 失敗するテストケースの作成
├── テストケース設計レビュー
└── 🔴 必須チェックポイント: テスト失敗の確認

### 0:20-0:50 [Green Phase - 最小実装]
├── 全チーム: テストを通す最小限の実装
├── 実装コードレビュー
└── 🟢 必須チェックポイント: 全テスト成功の確認

### 0:50-0:90 [Refactor Phase - 品質向上]
├── 全チーム: コード品質改善・最適化
├── テストが通り続けることの確認
└── 🔵 必須チェックポイント: リファクタリング後のテスト成功確認
```

### 従来型チーム構成からの移行
既存の「実装チーム」「テストチーム」構成からTDD体制への移行方法：

1. **チーム統合**: 実装担当とテスト担当を同一チームに統合
2. **ペアプログラミング**: テスト作成者と実装者のペア作業
3. **役割ローテーション**: Red/Green/Refactor各フェーズで役割交代

## 📊 継続的改善

### テスト戦略の継続的改善
1. **Phase毎効果測定**: テスト品質・バグ発見率・開発効率
2. **手法調整**: Red-Green-Refactorサイクルの最適化
3. **ツール改善**: より効率的なテストツール・環境の検討

### Phase A3での重点改善事項
- WebApplicationFactory統合テストの完全動作確認
- ASP.NET Core Identity統合テストの標準化
- Blazor Server UIテストの体系化

---

**注意事項**:
- このガイドは全Phase共通で使用
- Phase特性に応じたテスト重点の調整は各Phaseで実施
- 継続的改善により本ガイド自体も更新対象
- 運用手順は `/Doc/08_Organization/Rules/組織管理運用マニュアル.md` を参照