# ãƒ†ã‚¹ãƒˆæˆ¦ç•¥ã‚¬ã‚¤ãƒ‰ï¼ˆãƒ†ã‚¹ãƒˆãƒ•ã‚¡ãƒ¼ã‚¹ãƒˆé–‹ç™ºï¼‰

**ç›®çš„**: Phase A2ã§ç™ºç”Ÿã—ãŸçµ±åˆãƒ†ã‚¹ãƒˆèª²é¡Œã®æ ¹æœ¬çš„è§£æ±ºãƒ»å“è³ªå‘ä¸Šãƒ»Claudeä¸»ä½“é–‹ç™ºã®æœ€é©åŒ–  
**é©ç”¨ç¯„å›²**: å…¨Phaseãƒ»å…¨Step  
**ä½œæˆæ—¥**: 2025-07-21  
**æœ€çµ‚æ›´æ–°**: 2025-07-21  

## ğŸ”´ ãƒ†ã‚¹ãƒˆãƒ•ã‚¡ãƒ¼ã‚¹ãƒˆé–‹ç™ºã®å¿…é ˆåŒ–

### åŸºæœ¬æ–¹é‡
**ç›®çš„**: Phase A2ã§ç™ºç”Ÿã—ãŸçµ±åˆãƒ†ã‚¹ãƒˆèª²é¡Œã®æ ¹æœ¬çš„è§£æ±ºãƒ»å“è³ªå‘ä¸Šãƒ»Claudeä¸»ä½“é–‹ç™ºã®æœ€é©åŒ–

### ğŸ”´ å¿…é ˆå®Ÿè¡Œãƒ«ãƒ¼ãƒ«
```
å„Stepå®Ÿè£…æ™‚ã«ä»¥ä¸‹ã‚’å¿…é ˆå®Ÿè¡Œ:
â–¡ ãƒ†ã‚¹ãƒˆã‚±ãƒ¼ã‚¹äº‹å‰è¨­è¨ˆãƒ»å®Ÿè£…
â–¡ å®Ÿè£…å‰ã®ãƒ†ã‚¹ãƒˆå¤±æ•—ç¢ºèªï¼ˆRedï¼‰
â–¡ æœ€å°é™å®Ÿè£…ã«ã‚ˆã‚‹ãƒ†ã‚¹ãƒˆæˆåŠŸï¼ˆGreenï¼‰
â–¡ ãƒªãƒ•ã‚¡ã‚¯ã‚¿ãƒªãƒ³ã‚°ãƒ»å“è³ªå‘ä¸Šï¼ˆRefactorï¼‰
â–¡ ãƒ†ã‚¹ãƒˆã‚«ãƒãƒ¬ãƒƒã‚¸80%ä»¥ä¸Šç¶­æŒ
```

### ä»•æ§˜ãƒ™ãƒ¼ã‚¹Red-Green-Refactorã‚µã‚¤ã‚¯ãƒ«
1. **ä»•æ§˜ç¢ºèª**: æ©Ÿèƒ½ä»•æ§˜æ›¸ã®è©²å½“ç®‡æ‰€ã‚’ç²¾èª­ã—ã€ãƒ†ã‚¹ãƒˆã™ã¹ãå†…å®¹ã‚’ç‰¹å®š
2. **Red**: ä»•æ§˜ã«åŸºã¥ã„ã¦å¤±æ•—ã™ã‚‹ãƒ†ã‚¹ãƒˆã‚’æ›¸ãï¼ˆè‚¯å®šçš„ä»•æ§˜ãƒ»å¦å®šçš„ä»•æ§˜ä¸¡æ–¹ï¼‰
3. **Green**: ä»•æ§˜ã‚’æº€ãŸã™æœ€å°é™ã®ã‚³ãƒ¼ãƒ‰ã‚’æ›¸ã
4. **Refactor**: ã‚³ãƒ¼ãƒ‰ã‚’æ”¹å–„ã—ã€ä»•æ§˜æº–æ‹ ã¨ãƒ†ã‚¹ãƒˆæˆåŠŸã‚’ç¶­æŒ

## ğŸ“Š ãƒ†ã‚¹ãƒˆå±¤åˆ¥æˆ¦ç•¥

### ğŸŸ¢ å˜ä½“ãƒ†ã‚¹ãƒˆï¼ˆå„Stepå®Ÿè£…ã¨åŒæ™‚ï¼‰

**å¯¾è±¡**: F# Domainãƒ»Applicationå±¤ã€C# Infrastructureãƒ»Webå±¤

**æ‰‹æ³•**: ãƒ†ã‚¹ãƒˆãƒ•ã‚¡ãƒ¼ã‚¹ãƒˆï¼ˆRedâ†’Greenâ†’Refactorï¼‰

**å“è³ªåŸºæº–**: 
- å„ãƒ¡ã‚½ãƒƒãƒ‰ãƒ»é–¢æ•°ã®ãƒ†ã‚¹ãƒˆã‚±ãƒ¼ã‚¹ä½œæˆ
- ã‚¨ãƒƒã‚¸ã‚±ãƒ¼ã‚¹ãƒ»ç•°å¸¸ç³»ãƒ†ã‚¹ãƒˆã®åŒ…æ‹¬çš„å®Ÿè£…
- F# FsUnitã€C# xUnitæ´»ç”¨

**ä»•æ§˜æº–æ‹ å®Ÿè£…ãƒ‘ã‚¿ãƒ¼ãƒ³**:
```fsharp
// F# Domainå±¤ãƒ†ã‚¹ãƒˆä¾‹
// ä»•æ§˜æ›¸3.2.1æº–æ‹ : ãƒ¡ãƒ¼ãƒ«ã‚¢ãƒ‰ãƒ¬ã‚¹å½¢å¼æ¤œè¨¼
[<Test>]
let ``Email.create should validate correct email format per spec 3.2.1`` () =
    // Arrange
    let validEmail = "test@example.com"
    
    // Act & Assert
    match Email.create validEmail with
    | Ok email -> email.Value |> should equal validEmail
    | Error _ -> failwith "Valid email should not fail"

// ä»•æ§˜æ›¸3.2.1æº–æ‹ : ç„¡åŠ¹ãªãƒ¡ãƒ¼ãƒ«ã‚¢ãƒ‰ãƒ¬ã‚¹æ‹’å¦
[<Test>]
let ``Email.create should reject invalid format per spec 3.2.1`` () =
    // Arrange
    let invalidEmail = "invalid-email"
    
    // Act & Assert
    match Email.create invalidEmail with
    | Ok _ -> failwith "Invalid email should fail validation"
    | Error error -> error |> should contain "Invalid email format"
```

**å¦å®šçš„ä»•æ§˜ãƒ†ã‚¹ãƒˆãƒ‘ã‚¿ãƒ¼ãƒ³**:
```fsharp
// ä»•æ§˜æ›¸2.1.1æº–æ‹ : ãƒ­ã‚°ã‚¤ãƒ³å¤±æ•—ã«ã‚ˆã‚‹ãƒ­ãƒƒã‚¯ã‚¢ã‚¦ãƒˆæ©Ÿæ§‹ã¯è¨­ã‘ãªã„
[<Test>]
let ``Login should not lock account after multiple failures per spec 2.1.1`` () =
    // Arrange - 10å›ã®é€£ç¶šãƒ­ã‚°ã‚¤ãƒ³å¤±æ•—ã‚’æº–å‚™
    let loginAttempts = [1..10]
    let mutable finalResult = None
    
    // Act - 10å›å¤±æ•—å¾Œã‚‚æ­£ã—ã„ãƒ‘ã‚¹ãƒ¯ãƒ¼ãƒ‰ã§ãƒ­ã‚°ã‚¤ãƒ³å¯èƒ½
    for _ in loginAttempts do
        AuthService.login "user@example.com" "wrongpassword" |> ignore
    
    finalResult <- Some (AuthService.login "user@example.com" "correctpassword")
    
    // Assert - ãƒ­ãƒƒã‚¯ã‚¢ã‚¦ãƒˆã•ã‚Œã¦ã„ãªã„ã“ã¨ã‚’ç¢ºèª
    match finalResult with
    | Some (Ok _) -> () // æˆåŠŸï¼šä»•æ§˜æº–æ‹ 
    | Some (Error msg) when msg.Contains("locked") -> 
        failwith "Account should not be locked per spec 2.1.1"
    | _ -> failwith "Unexpected result"
```

```csharp
// C# Infrastructureå±¤ãƒ†ã‚¹ãƒˆä¾‹
[Fact]
public async Task GetByIdAsync_ShouldReturnUser_WhenUserExists()
{
    // Arrange
    var userId = Guid.NewGuid();
    var expectedUser = CreateTestUser(userId);
    
    // Act
    var result = await _userRepository.GetByIdAsync(userId);
    
    // Assert
    Assert.NotNull(result);
    Assert.Equal(expectedUser.Id, result.Id);
}
```

### ğŸŸ¡ çµ±åˆãƒ†ã‚¹ãƒˆï¼ˆãƒ¬ã‚¤ãƒ¤ãƒ¼çµ±åˆæ™‚ï¼‰

**å¯¾è±¡**: F#â†”C#å¢ƒç•Œãƒ»DBæ¥ç¶šãƒ»ASP.NET Core Identityé€£æº

**æ‰‹æ³•**: WebApplicationFactoryãƒ»In-Memory Databaseæ´»ç”¨

**å“è³ªåŸºæº–**:
- å±¤é–“ãƒ‡ãƒ¼ã‚¿ãƒ•ãƒ­ãƒ¼å®Œå…¨æ¤œè¨¼
- Phase A2ã§ç™ºç”Ÿã—ãŸHTTP 500ã‚¨ãƒ©ãƒ¼é¡ã®äº‹å‰ç™ºè¦‹ãƒ»è§£æ±º
- èªè¨¼ãƒ»èªå¯ãƒ•ãƒ­ãƒ¼å®Œå…¨æ¤œè¨¼

**å®Ÿè£…ãƒ‘ã‚¿ãƒ¼ãƒ³**:
```csharp
public class UserIntegrationTests : IClassFixture<WebApplicationFactory<Program>>
{
    private readonly WebApplicationFactory<Program> _factory;
    
    public UserIntegrationTests(WebApplicationFactory<Program> factory)
    {
        _factory = factory;
    }
    
    [Fact]
    public async Task CreateUser_ShouldReturnSuccess_WhenValidData()
    {
        // Arrange
        var client = _factory.CreateClient();
        var createUserDto = new CreateUserDto { /* ... */ };
        
        // Act
        var response = await client.PostAsJsonAsync("/api/users", createUserDto);
        
        // Assert
        response.EnsureSuccessStatusCode();
        var createdUser = await response.Content.ReadFromJsonAsync<UserDto>();
        Assert.NotNull(createdUser);
    }
}
```

### Phase A2èª²é¡Œã®å…·ä½“çš„è§£æ±º
**èª²é¡Œ**: WebApplicationFactoryçµ±åˆãƒ†ã‚¹ãƒˆã§HTTP 500ã‚¨ãƒ©ãƒ¼ï¼ˆ6ãƒ†ã‚¹ãƒˆå¤±æ•—ï¼‰

**è§£æ±ºç­–**:
1. **ASP.NET Core Identity In-Memory DBå¯¾å¿œè¨­å®š**
2. **TestInitialDataServiceå®Œå…¨å®Ÿè£…**
3. **UserManager/RoleManager/AuthenticationStateProvideré©åˆ‡ãªãƒ¢ãƒƒã‚¯**
4. **DbContextä¾å­˜é–¢ä¿‚æ•´ç†ãƒ»é‡è¤‡è§£æ¶ˆ**

### ğŸ”´ E2Eãƒ†ã‚¹ãƒˆï¼ˆUIå®Ÿè£…å®Œäº†æ™‚ï¼‰

**å¯¾è±¡**: Blazor Server UIãƒ»ãƒ¦ãƒ¼ã‚¶ãƒ¼ã‚·ãƒŠãƒªã‚ªãƒ»ãƒ–ãƒ©ã‚¦ã‚¶å‹•ä½œ

**æ‰‹æ³•**: Playwrightãƒ»Seleniumæ´»ç”¨æ¤œè¨

**å“è³ªåŸºæº–**:
- ä¸»è¦ãƒ¦ãƒ¼ã‚¶ãƒ¼ã‚·ãƒŠãƒªã‚ªã®è‡ªå‹•åŒ–
- ãƒ–ãƒ©ã‚¦ã‚¶äº’æ›æ€§ç¢ºèª
- ãƒ¬ã‚¹ãƒãƒ³ã‚·ãƒ–ãƒ‡ã‚¶ã‚¤ãƒ³æ¤œè¨¼

**å®Ÿè£…ãƒ‘ã‚¿ãƒ¼ãƒ³**:
```csharp
[Test]
public async Task UserLogin_ShouldNavigateToHome_WhenCredentialsValid()
{
    // Arrange
    await using var browser = await Playwright.Chromium.LaunchAsync();
    var page = await browser.NewPageAsync();
    
    // Act
    await page.GotoAsync("https://localhost:5001/login");
    await page.FillAsync("#email", "test@example.com");
    await page.FillAsync("#password", "TestPassword123!");
    await page.ClickAsync("#login-button");
    
    // Assert
    await page.WaitForURLAsync("**/home");
    var title = await page.TextContentAsync("h1");
    Assert.Contains("Welcome", title);
}
```

### ğŸŸ£ å—ã‘å…¥ã‚Œãƒ†ã‚¹ãƒˆï¼ˆPhaseå®Œäº†å‰ï¼‰

**å¯¾è±¡**: ãƒ“ã‚¸ãƒã‚¹è¦ä»¶ãƒ»ãƒ¦ãƒ¼ã‚¶ãƒ“ãƒªãƒ†ã‚£ãƒ»ãƒ‘ãƒ•ã‚©ãƒ¼ãƒãƒ³ã‚¹

**æ‰‹æ³•**: æ‰‹å‹•ãƒ†ã‚¹ãƒˆãƒ»å“è³ªãƒã‚§ãƒƒã‚¯ãƒªã‚¹ãƒˆ

**å“è³ªåŸºæº–**:
- è¦ä»¶ä»•æ§˜æ›¸ã¨ã®å®Œå…¨æº–æ‹ ç¢ºèª
- ãƒ¦ãƒ¼ã‚¶ãƒ“ãƒªãƒ†ã‚£åŸºæº–ã‚¯ãƒªã‚¢
- ãƒ‘ãƒ•ã‚©ãƒ¼ãƒãƒ³ã‚¹åŸºæº–æº€è¶³

## ğŸ”„ Stepå®Ÿè¡Œæ™‚ãƒ†ã‚¹ãƒˆçµ±åˆ

### Stepé–‹å§‹æ™‚ï¼ˆãƒ†ã‚¹ãƒˆãƒ•ã‚¡ãƒ¼ã‚¹ãƒˆæº–å‚™ï¼‰
1. **ãƒ†ã‚¹ãƒˆè¨ˆç”»ç­–å®š**: Stepç›®æ¨™ã«åŸºã¥ããƒ†ã‚¹ãƒˆã‚±ãƒ¼ã‚¹è¨­è¨ˆ
2. **ãƒ†ã‚¹ãƒˆç’°å¢ƒæº–å‚™**: å¿…è¦ãªãƒ†ã‚¹ãƒˆãƒ‡ãƒ¼ã‚¿ãƒ»ãƒ¢ãƒƒã‚¯ãƒ»è¨­å®šæº–å‚™
3. **ãƒ†ã‚¹ãƒˆå®Ÿè£…é–‹å§‹**: æ©Ÿèƒ½å®Ÿè£…å‰ã®ãƒ†ã‚¹ãƒˆã‚±ãƒ¼ã‚¹ä½œæˆ

### Stepå®Ÿè¡Œä¸­ï¼ˆRed-Green-Refactorã‚µã‚¤ã‚¯ãƒ«ï¼‰
1. **Red**: å¤±æ•—ã™ã‚‹ãƒ†ã‚¹ãƒˆã®å®Ÿè£…ãƒ»ç¢ºèª
2. **Green**: æœ€å°é™å®Ÿè£…ã«ã‚ˆã‚‹ãƒ†ã‚¹ãƒˆæˆåŠŸ
3. **Refactor**: ã‚³ãƒ¼ãƒ‰å“è³ªå‘ä¸Šãƒ»ãƒ†ã‚¹ãƒˆæˆåŠŸç¶­æŒ

### Stepçµ‚äº†æ™‚ãƒ†ã‚¹ãƒˆå“è³ªç¢ºèª

**ğŸ”´ Stepçµ‚äº†æ™‚å¿…é ˆç¢ºèª**:
```
â–¡ æ–°è¦å®Ÿè£…æ©Ÿèƒ½ã®å˜ä½“ãƒ†ã‚¹ãƒˆå®Œæˆãƒ»æˆåŠŸ
â–¡ çµ±åˆãƒ†ã‚¹ãƒˆå®Ÿè¡Œãƒ»èª²é¡Œè§£æ±ºå®Œäº†
â–¡ ãƒ†ã‚¹ãƒˆã‚«ãƒãƒ¬ãƒƒã‚¸80%ä»¥ä¸Šç¶­æŒç¢ºèª
â–¡ å…¨ãƒ†ã‚¹ãƒˆæˆåŠŸçŠ¶æ…‹ã§ã®Stepå®Œäº†
â–¡ æ¬¡Stepãƒ†ã‚¹ãƒˆç’°å¢ƒãƒ»å‰ææ¡ä»¶ç¢ºèª
```

## ğŸ“‹ æŠ€è¡“ã‚¹ã‚¿ãƒƒã‚¯åˆ¥å®Ÿè£…ã‚¬ã‚¤ãƒ‰

### F# Domainãƒ»Applicationå±¤
- **ãƒ†ã‚¹ãƒˆãƒ•ãƒ¬ãƒ¼ãƒ ãƒ¯ãƒ¼ã‚¯**: NUnit + FsUnit
- **ãƒ¢ãƒƒã‚­ãƒ³ã‚°**: Moq for F#
- **é‡ç‚¹**: ç´”ç²‹é–¢æ•°ãƒ»å‹å®‰å…¨æ€§ãƒ»ãƒ“ã‚¸ãƒã‚¹ãƒ«ãƒ¼ãƒ«

### C# Infrastructureãƒ»Webå±¤
- **ãƒ†ã‚¹ãƒˆãƒ•ãƒ¬ãƒ¼ãƒ ãƒ¯ãƒ¼ã‚¯**: xUnit
- **çµ±åˆãƒ†ã‚¹ãƒˆ**: WebApplicationFactory
- **ãƒ¢ãƒƒã‚­ãƒ³ã‚°**: Moq
- **é‡ç‚¹**: DBæ“ä½œãƒ»HTTPé€šä¿¡ãƒ»èªè¨¼èªå¯

### Blazor Server UI
- **ã‚³ãƒ³ãƒãƒ¼ãƒãƒ³ãƒˆãƒ†ã‚¹ãƒˆ**: bUnit
- **E2Eãƒ†ã‚¹ãƒˆ**: Playwright.NET
- **é‡ç‚¹**: ãƒ¦ãƒ¼ã‚¶ãƒ¼ã‚¤ãƒ³ã‚¿ãƒ©ã‚¯ã‚·ãƒ§ãƒ³ãƒ»ãƒ¬ã‚¹ãƒãƒ³ã‚·ãƒ–å¯¾å¿œ

## ğŸ¯ å“è³ªä¿è¨¼æŒ‡æ¨™

### ãƒ†ã‚¹ãƒˆã‚«ãƒãƒ¬ãƒƒã‚¸ç›®æ¨™
- **å˜ä½“ãƒ†ã‚¹ãƒˆ**: 80%ä»¥ä¸Š
- **çµ±åˆãƒ†ã‚¹ãƒˆ**: ä¸»è¦ãƒ•ãƒ­ãƒ¼100%
- **E2Eãƒ†ã‚¹ãƒˆ**: é‡è¦ãƒ¦ãƒ¼ã‚¶ãƒ¼ã‚·ãƒŠãƒªã‚ª100%

### ãƒ‘ãƒ•ã‚©ãƒ¼ãƒãƒ³ã‚¹åŸºæº–
- **å˜ä½“ãƒ†ã‚¹ãƒˆå®Ÿè¡Œæ™‚é–“**: 10ç§’ä»¥å†…ï¼ˆå…¨ãƒ†ã‚¹ãƒˆï¼‰
- **çµ±åˆãƒ†ã‚¹ãƒˆå®Ÿè¡Œæ™‚é–“**: 60ç§’ä»¥å†…ï¼ˆå…¨ãƒ†ã‚¹ãƒˆï¼‰
- **E2Eãƒ†ã‚¹ãƒˆå®Ÿè¡Œæ™‚é–“**: 5åˆ†ä»¥å†…ï¼ˆå…¨ã‚·ãƒŠãƒªã‚ªï¼‰

### å“è³ªã‚²ãƒ¼ãƒˆ
- **å…¨ãƒ†ã‚¹ãƒˆæˆåŠŸ**: Stepå®Œäº†ã®å¿…é ˆæ¡ä»¶
- **ã‚«ãƒãƒ¬ãƒƒã‚¸ç¶­æŒ**: 80%ä»¥ä¸Šç¶­æŒå¿…é ˆ
- **ãƒ‘ãƒ•ã‚©ãƒ¼ãƒãƒ³ã‚¹åŠ£åŒ–**: å‰å›æ¯”10%ä»¥å†…

## ğŸ”§ å®Ÿè£…æ”¯æ´ãƒ„ãƒ¼ãƒ«

### æ¨å¥¨ãƒ„ãƒ¼ãƒ«
- **ãƒ†ã‚¹ãƒˆãƒ©ãƒ³ãƒŠãƒ¼**: dotnet test
- **ã‚«ãƒãƒ¬ãƒƒã‚¸**: coverlet.collector
- **ãƒ¬ãƒãƒ¼ãƒˆ**: ReportGenerator
- **CIçµ±åˆ**: GitHub Actions

### é–‹ç™ºç’°å¢ƒçµ±åˆ
```json
// .vscode/settings.json
{
    "dotnet.test.explorer.enabled": true,
    "dotnet.test.explorer.autoRun": true,
    "coverage-gutters.enableFileOnSave": true
}
```

## ğŸ¢ TDDå®Ÿè·µã®ãŸã‚ã®çµ„ç¹”è¨­è¨ˆãƒ†ãƒ³ãƒ—ãƒ¬ãƒ¼ãƒˆ

### ãƒ•ã‚£ãƒ¼ãƒãƒ£ãƒ¼ãƒãƒ¼ãƒ åˆ¶ï¼ˆæ¨å¥¨ï¼‰
å®Ÿè£…ã¨ãƒ†ã‚¹ãƒˆã‚’åˆ†é›¢ã›ãšã€æ©Ÿèƒ½å˜ä½ã§ãƒãƒ¼ãƒ ã‚’ç·¨æˆã—ã€å„ãƒãƒ¼ãƒ ãŒRed-Green-Refactorã‚µã‚¤ã‚¯ãƒ«ã‚’å®Œçµã§ãã‚‹ä½“åˆ¶ã€‚

```markdown
#### Feature Team 1: [æ©Ÿèƒ½å]
**è²¬ä»»ç¯„å›²**: [æ©Ÿèƒ½ã®è©³ç´°èª¬æ˜]
**TDDå®Ÿè·µè¨ˆç”»**:
- Red Phase: [ä½œæˆã™ã‚‹ãƒ†ã‚¹ãƒˆã‚±ãƒ¼ã‚¹ä¸€è¦§]
- Green Phase: [å®Ÿè£…ã™ã‚‹æœ€å°æ©Ÿèƒ½ä¸€è¦§]
- Refactor Phase: [æ”¹å–„äºˆå®šé …ç›®ä¸€è¦§]
**æŠ€è¡“ã‚¹ã‚¿ãƒƒã‚¯**: [ä½¿ç”¨ã™ã‚‹æŠ€è¡“ãƒ»ãƒ•ãƒ¬ãƒ¼ãƒ ãƒ¯ãƒ¼ã‚¯]
```

### TDDã‚µã‚¤ã‚¯ãƒ«å®Ÿè¡Œè¨ˆç”»ãƒ†ãƒ³ãƒ—ãƒ¬ãƒ¼ãƒˆ
```markdown
## TDDã‚µã‚¤ã‚¯ãƒ«å®Ÿè¡Œè¨ˆç”»
### 0:00-0:20 [Red Phase - ãƒ†ã‚¹ãƒˆä½œæˆ]
â”œâ”€â”€ å…¨ãƒãƒ¼ãƒ : å¤±æ•—ã™ã‚‹ãƒ†ã‚¹ãƒˆã‚±ãƒ¼ã‚¹ã®ä½œæˆ
â”œâ”€â”€ ãƒ†ã‚¹ãƒˆã‚±ãƒ¼ã‚¹è¨­è¨ˆãƒ¬ãƒ“ãƒ¥ãƒ¼
â””â”€â”€ ğŸ”´ å¿…é ˆãƒã‚§ãƒƒã‚¯ãƒã‚¤ãƒ³ãƒˆ: ãƒ†ã‚¹ãƒˆå¤±æ•—ã®ç¢ºèª

### 0:20-0:50 [Green Phase - æœ€å°å®Ÿè£…]
â”œâ”€â”€ å…¨ãƒãƒ¼ãƒ : ãƒ†ã‚¹ãƒˆã‚’é€šã™æœ€å°é™ã®å®Ÿè£…
â”œâ”€â”€ å®Ÿè£…ã‚³ãƒ¼ãƒ‰ãƒ¬ãƒ“ãƒ¥ãƒ¼
â””â”€â”€ ğŸŸ¢ å¿…é ˆãƒã‚§ãƒƒã‚¯ãƒã‚¤ãƒ³ãƒˆ: å…¨ãƒ†ã‚¹ãƒˆæˆåŠŸã®ç¢ºèª

### 0:50-0:90 [Refactor Phase - å“è³ªå‘ä¸Š]
â”œâ”€â”€ å…¨ãƒãƒ¼ãƒ : ã‚³ãƒ¼ãƒ‰å“è³ªæ”¹å–„ãƒ»æœ€é©åŒ–
â”œâ”€â”€ ãƒ†ã‚¹ãƒˆãŒé€šã‚Šç¶šã‘ã‚‹ã“ã¨ã®ç¢ºèª
â””â”€â”€ ğŸ”µ å¿…é ˆãƒã‚§ãƒƒã‚¯ãƒã‚¤ãƒ³ãƒˆ: ãƒªãƒ•ã‚¡ã‚¯ã‚¿ãƒªãƒ³ã‚°å¾Œã®ãƒ†ã‚¹ãƒˆæˆåŠŸç¢ºèª
```

### å¾“æ¥å‹ãƒãƒ¼ãƒ æ§‹æˆã‹ã‚‰ã®ç§»è¡Œ
æ—¢å­˜ã®ã€Œå®Ÿè£…ãƒãƒ¼ãƒ ã€ã€Œãƒ†ã‚¹ãƒˆãƒãƒ¼ãƒ ã€æ§‹æˆã‹ã‚‰TDDä½“åˆ¶ã¸ã®ç§»è¡Œæ–¹æ³•ï¼š

1. **ãƒãƒ¼ãƒ çµ±åˆ**: å®Ÿè£…æ‹…å½“ã¨ãƒ†ã‚¹ãƒˆæ‹…å½“ã‚’åŒä¸€ãƒãƒ¼ãƒ ã«çµ±åˆ
2. **ãƒšã‚¢ãƒ—ãƒ­ã‚°ãƒ©ãƒŸãƒ³ã‚°**: ãƒ†ã‚¹ãƒˆä½œæˆè€…ã¨å®Ÿè£…è€…ã®ãƒšã‚¢ä½œæ¥­
3. **å½¹å‰²ãƒ­ãƒ¼ãƒ†ãƒ¼ã‚·ãƒ§ãƒ³**: Red/Green/Refactorå„ãƒ•ã‚§ãƒ¼ã‚ºã§å½¹å‰²äº¤ä»£

## ğŸ“‹ è©³ç´°å®Ÿè£…ã‚µãƒ³ãƒ—ãƒ«ã‚³ãƒ¼ãƒ‰ï¼ˆADR_009è©³ç´°ç‰ˆï¼‰

### F# Domainå±¤å˜ä½“ãƒ†ã‚¹ãƒˆä¾‹

```fsharp
// âœ… Domainå±¤å˜ä½“ãƒ†ã‚¹ãƒˆï¼ˆF#ï¼‰- æœ€é«˜å„ªå…ˆåº¦
module UbiquitousLangDomainTests =
    open Xunit
    open FsUnit.Xunit
    
    [<Fact>]
    let ``ç©ºã®ãƒ¦ãƒ“ã‚­ã‚¿ã‚¹è¨€èªåã§ãƒãƒªãƒ‡ãƒ¼ã‚·ãƒ§ãƒ³ã‚¨ãƒ©ãƒ¼ãŒè¿”ã•ã‚Œã‚‹`` () =
        // ğŸ¯ F#ç´”ç²‹é–¢æ•°ãƒ†ã‚¹ãƒˆ: å¤–éƒ¨ä¾å­˜ãªã—ã€é«˜é€Ÿå®Ÿè¡Œ
        let command = { Name = ""; ProjectId = 1; Description = "test" }
        
        let result = UbiquitousLangDomain.validateUbiquitousLang command
        
        // âœ… FsUnit: F#ã«æœ€é©åŒ–ã•ã‚ŒãŸã‚¢ã‚µãƒ¼ã‚·ãƒ§ãƒ³
        result |> should equal (Error EmptyUbiquitousLangName)
    
    [<Theory>]
    [<InlineData(Draft, UnderReview, true)>]
    [<InlineData(UnderReview, Approved, true)>]
    [<InlineData(UnderReview, Rejected, true)>]
    [<InlineData(Rejected, Draft, true)>]
    [<InlineData(Draft, Approved, false)>]
    [<InlineData(Approved, Draft, false)>]
    let ``çŠ¶æ…‹é·ç§»ãƒãƒªãƒ‡ãƒ¼ã‚·ãƒ§ãƒ³`` (current: UbiquitousLangStatus) (next: UbiquitousLangStatus) (shouldSucceed: bool) =
        // ğŸ­ F#ãƒ‘ã‚¿ãƒ¼ãƒ³ãƒãƒƒãƒãƒ³ã‚°: å…¨ã‚±ãƒ¼ã‚¹ç¶²ç¾…ãƒ†ã‚¹ãƒˆ
        let result = UbiquitousLangDomain.validateStatusTransition current next
        
        match shouldSucceed with
        | true -> result |> should equal (Ok ())
        | false -> 
            match result with
            | Error (InvalidStatusTransition _) -> Assert.True(true)
            | _ -> Assert.True(false, $"ç„¡åŠ¹ãªçŠ¶æ…‹é·ç§»ã‚¨ãƒ©ãƒ¼ãŒæœŸå¾…ã•ã‚Œã‚‹: {current} -> {next}")
```

### C# Applicationå±¤å˜ä½“ãƒ†ã‚¹ãƒˆä¾‹

```csharp
// âœ… Applicationå±¤å˜ä½“ãƒ†ã‚¹ãƒˆï¼ˆC#ï¼‰
public class UbiquitousLangApplicationServiceTests
{
    private readonly IUbiquitousLangRepository _mockRepository;
    private readonly ILogger<UbiquitousLangApplicationService> _mockLogger;
    private readonly UbiquitousLangApplicationService _service;
    
    public UbiquitousLangApplicationServiceTests()
    {
        // ğŸ”§ NSubstitute: ãƒ¢ãƒƒã‚¯ãƒ»ã‚¹ã‚¿ãƒ–ä½œæˆ
        _mockRepository = Substitute.For<IUbiquitousLangRepository>();
        _mockLogger = Substitute.For<ILogger<UbiquitousLangApplicationService>>();
        _service = new UbiquitousLangApplicationService(_mockLogger, _mockRepository);
    }
    
    [Fact]
    public async Task CreateUbiquitousLangAsync_æ­£å¸¸ãªå…¥åŠ›ã§æˆåŠŸ()
    {
        // Arrange
        var command = new CreateUbiquitousLangCommand 
        { 
            Name = "ãƒ¦ãƒ“ã‚­ã‚¿ã‚¹è¨€èª", 
            ProjectId = 1, 
            Description = "èª¬æ˜" 
        };
        
        _mockRepository.CreateAsync(Arg.Any<UbiquitousLang>())
            .Returns(Task.FromResult(new UbiquitousLang { Id = 1, Name = "ãƒ¦ãƒ“ã‚­ã‚¿ã‚¹è¨€èª" }));
        
        // Act
        var result = await _service.CreateUbiquitousLangAsync(command);
        
        // Assert
        result.Should().NotBeNull();
        result.IsSuccess.Should().BeTrue();
        result.Value.Name.Should().Be("ãƒ¦ãƒ“ã‚­ã‚¿ã‚¹è¨€èª");
        
        // ğŸ” ãƒ¢ãƒƒã‚¯å‘¼ã³å‡ºã—ç¢ºèª
        await _mockRepository.Received(1).CreateAsync(Arg.Any<UbiquitousLang>());
    }
}
```

### E2Eãƒ†ã‚¹ãƒˆå®Ÿè£…ä¾‹

```csharp
// âœ… E2Eãƒ†ã‚¹ãƒˆ: ä¸»è¦ã‚·ãƒŠãƒªã‚ªã®ã¿
public class UbiquitousLangWorkflowE2ETests : IAsyncDisposable
{
    private IPlaywright _playwright;
    private IBrowser _browser;
    private IPage _page;
    
    [Fact]
    public async Task ãƒ¦ãƒ“ã‚­ã‚¿ã‚¹è¨€èªæ‰¿èªãƒ¯ãƒ¼ã‚¯ãƒ•ãƒ­ãƒ¼_æ­£å¸¸ç³»()
    {
        // ğŸ¬ ã‚»ãƒƒãƒˆã‚¢ãƒƒãƒ—
        _playwright = await Playwright.CreateAsync();
        _browser = await _playwright.Chromium.LaunchAsync(new BrowserTypeLaunchOptions { Headless = true });
        _page = await _browser.NewPageAsync();
        
        // ğŸ“ Step1: ãƒ­ã‚°ã‚¤ãƒ³
        await _page.GotoAsync("https://localhost:5001/login");
        await _page.FillAsync("#email", "admin@example.com");
        await _page.FillAsync("#password", "Password123!");
        await _page.ClickAsync("#login-button");
        
        // âœ… ãƒ­ã‚°ã‚¤ãƒ³æˆåŠŸç¢ºèª
        await _page.WaitForSelectorAsync("#dashboard");
        
        // ğŸ“ Step2: ãƒ¦ãƒ“ã‚­ã‚¿ã‚¹è¨€èªä½œæˆ
        await _page.ClickAsync("#ubiquitous-langs-menu");
        await _page.ClickAsync("#create-button");
        await _page.FillAsync("#name", "æ–°ã—ã„ãƒ¦ãƒ“ã‚­ã‚¿ã‚¹è¨€èª");
        await _page.FillAsync("#description", "ãƒ†ã‚¹ãƒˆç”¨ã®èª¬æ˜");
        await _page.ClickAsync("#submit-button");
        
        // âœ… ä½œæˆæˆåŠŸç¢ºèª
        await _page.WaitForSelectorAsync("#success-message");
        var successText = await _page.TextContentAsync("#success-message");
        successText.Should().Contain("æ­£å¸¸ã«ä½œæˆã•ã‚Œã¾ã—ãŸ");
    }
}
```

### ãƒ†ã‚¹ãƒˆãƒ—ãƒ­ã‚¸ã‚§ã‚¯ãƒˆè¨­å®š

```xml
<!-- âœ… ãƒ†ã‚¹ãƒˆãƒ—ãƒ­ã‚¸ã‚§ã‚¯ãƒˆå…±é€šãƒ‘ãƒƒã‚±ãƒ¼ã‚¸ -->
<PackageReference Include="Microsoft.NET.Test.Sdk" Version="17.8.0" />

<!-- å˜ä½“ãƒ»çµ±åˆãƒ†ã‚¹ãƒˆ -->
<PackageReference Include="xUnit" Version="2.4.2" />
<PackageReference Include="xunit.runner.visualstudio" Version="2.4.5" />

<!-- ã‚¢ã‚µãƒ¼ã‚·ãƒ§ãƒ³ -->
<PackageReference Include="FluentAssertions" Version="6.12.0" />
<PackageReference Include="FsUnit.xUnit" Version="5.6.0" /> <!-- F#å°‚ç”¨ -->

<!-- ãƒ¢ãƒƒã‚¯ -->
<PackageReference Include="NSubstitute" Version="5.1.0" />

<!-- E2Eãƒ†ã‚¹ãƒˆ -->
<PackageReference Include="Microsoft.Playwright" Version="1.40.0" />

<!-- ãƒ‡ãƒ¼ã‚¿ãƒ™ãƒ¼ã‚¹ãƒ†ã‚¹ãƒˆ -->
<PackageReference Include="Microsoft.EntityFrameworkCore.InMemory" Version="8.0.0" />
<PackageReference Include="Testcontainers.PostgreSql" Version="3.5.0" />
```

### ç¶™ç¶šçš„ã‚¤ãƒ³ãƒ†ã‚°ãƒ¬ãƒ¼ã‚·ãƒ§ãƒ³è¨­å®š

```yaml
# âœ… GitHub Actionsè¨­å®šä¾‹
name: Tests
on: [push, pull_request]

jobs:
  unit-tests:
    runs-on: ubuntu-latest
    steps:
    - uses: actions/checkout@v3
    - name: Setup .NET
      uses: actions/setup-dotnet@v3
      with:
        dotnet-version: '8.0.x'
    - name: Restore dependencies
      run: dotnet restore
    - name: Run unit tests
      run: dotnet test --filter Category=Unit --logger trx --results-directory TestResults
    - name: Upload test results
      uses: actions/upload-artifact@v3
      with:
        name: test-results
        path: TestResults
```

### DatabaseFixtureå®Ÿè£…ä¾‹

```csharp
// âœ… çµ±åˆãƒ†ã‚¹ãƒˆç”¨ãƒ‡ãƒ¼ã‚¿ãƒ™ãƒ¼ã‚¹è¨­å®š
public class DatabaseFixture : IAsyncDisposable
{
    public IDbContextFactory<UbiquitousLanguageDbContext> ContextFactory { get; private set; }
    
    public async Task InitializeAsync()
    {
        var services = new ServiceCollection();
        services.AddDbContextFactory<UbiquitousLanguageDbContext>(options =>
        {
            // ğŸ”§ In-Memory Database: é«˜é€Ÿãƒ»ç‹¬ç«‹ã—ãŸãƒ†ã‚¹ãƒˆå®Ÿè¡Œ
            options.UseInMemoryDatabase($"TestDb_{Guid.NewGuid()}");
        });
        
        var serviceProvider = services.BuildServiceProvider();
        ContextFactory = serviceProvider.GetRequiredService<IDbContextFactory<UbiquitousLanguageDbContext>>();
        
        using var context = ContextFactory.CreateDbContext();
        await context.Database.EnsureCreatedAsync();
    }
    
    public async Task DisposeAsync()
    {
        // ğŸ§¹ ãƒ†ã‚¹ãƒˆå¾Œã®ç¢ºå®Ÿãªã‚¯ãƒªãƒ¼ãƒ³ã‚¢ãƒƒãƒ—
        using var context = ContextFactory.CreateDbContext();
        await context.Database.EnsureDeletedAsync();
    }
}
```

## ğŸ“Š ç¶™ç¶šçš„æ”¹å–„

### ãƒ†ã‚¹ãƒˆæˆ¦ç•¥ã®ç¶™ç¶šçš„æ”¹å–„
1. **Phaseæ¯åŠ¹æœæ¸¬å®š**: ãƒ†ã‚¹ãƒˆå“è³ªãƒ»ãƒã‚°ç™ºè¦‹ç‡ãƒ»é–‹ç™ºåŠ¹ç‡
2. **æ‰‹æ³•èª¿æ•´**: Red-Green-Refactorã‚µã‚¤ã‚¯ãƒ«ã®æœ€é©åŒ–
3. **ãƒ„ãƒ¼ãƒ«æ”¹å–„**: ã‚ˆã‚ŠåŠ¹ç‡çš„ãªãƒ†ã‚¹ãƒˆãƒ„ãƒ¼ãƒ«ãƒ»ç’°å¢ƒã®æ¤œè¨

### Phase A3ã§ã®é‡ç‚¹æ”¹å–„äº‹é …
- WebApplicationFactoryçµ±åˆãƒ†ã‚¹ãƒˆã®å®Œå…¨å‹•ä½œç¢ºèª
- ASP.NET Core Identityçµ±åˆãƒ†ã‚¹ãƒˆã®æ¨™æº–åŒ–
- Blazor Server UIãƒ†ã‚¹ãƒˆã®ä½“ç³»åŒ–

---

**æ³¨æ„äº‹é …**:
- ã“ã®ã‚¬ã‚¤ãƒ‰ã¯å…¨Phaseå…±é€šã§ä½¿ç”¨
- Phaseç‰¹æ€§ã«å¿œã˜ãŸãƒ†ã‚¹ãƒˆé‡ç‚¹ã®èª¿æ•´ã¯å„Phaseã§å®Ÿæ–½
- è©³ç´°ã‚µãƒ³ãƒ—ãƒ«ã‚³ãƒ¼ãƒ‰ã¯2025-07-27ã«ADR_009ã‹ã‚‰ç§»ç®¡ãƒ»çµ±åˆ
- ç¶™ç¶šçš„æ”¹å–„ã«ã‚ˆã‚Šæœ¬ã‚¬ã‚¤ãƒ‰è‡ªä½“ã‚‚æ›´æ–°å¯¾è±¡
- é‹ç”¨æ‰‹é †ã¯ `/Doc/08_Organization/Rules/çµ„ç¹”ç®¡ç†é‹ç”¨ãƒãƒ‹ãƒ¥ã‚¢ãƒ«.md` ã‚’å‚ç…§