# 仕様準拠ガイド

**目的**: 仕様書逸脱を防止し、仕様準拠を保証する開発プロセスの確立  
**適用範囲**: 全Phase・全Step  
**作成日**: 2025-07-23  
**最終更新**: 2025-07-23  

## 🎯 基本原則

### 仕様ファーストの開発思想
**すべての実装は仕様書に基づく**
- コードよりも先に仕様を理解する
- テストよりも先に仕様を確認する
- 実装判断に迷ったら仕様書に立ち返る

### 仕様準拠の3つの柱
1. **事前確認**: 実装前の仕様精読と理解
2. **実装時検証**: 仕様に基づくテスト設計と実装
3. **事後監査**: 仕様準拠の継続的確認

## 📋 仕様書参照の必須化

### Step開始時の仕様確認プロセス
```
1. 該当機能の仕様書箇所を特定
   └─ 機能仕様書の目次から該当セクションを探索
   └─ 関連する複数セクションをリストアップ

2. 仕様精読と理解度確認
   └─ 肯定的仕様（実装すべき機能）の抽出
   └─ 否定的仕様（実装してはいけない機能）の抽出
   └─ ビジネスルールの完全理解

3. 仕様準拠チェックリスト作成
   └─ 実装すべき機能リスト
   └─ 実装してはいけない機能リスト
   └─ 検証すべきビジネスルール
```

### 仕様書項番の記録ルール
**すべての実装に仕様書項番を紐付ける**

```csharp
// 仕様書2.1.1準拠: ログイン機能
// ビジネスルール: ログイン失敗によるロックアウト機構は設けない
public async Task<IActionResult> Login(LoginModel model)
{
    // 実装内容
}
```

```fsharp
// 仕様書3.2.1準拠: プロジェクト作成機能
// 必須項目: プロジェクト名、説明
let createProject (name: ProjectName) (description: Description) =
    // 実装内容
```

## 🚫 否定的仕様（〜しない）への対応

### 否定的仕様の種類と対応方法

#### 1. 機能制限型
**例**: 「ログイン失敗によるロックアウト機構は設けない」

**テスト手法**:
```csharp
[Fact]
public async Task Login_ShouldNotLockAccount_EvenAfterMultipleFailures()
{
    // Arrange
    var loginAttempts = 10; // 通常のロックアウト閾値を超える回数
    
    // Act
    for (int i = 0; i < loginAttempts; i++)
    {
        await _authService.Login("user@example.com", "wrongpassword");
    }
    
    // Assert - 正しいパスワードでログイン可能であることを確認
    var result = await _authService.Login("user@example.com", "correctpassword");
    Assert.True(result.Succeeded);
    Assert.False(result.IsLockedOut); // ロックアウトされていないことを確認
}
```

#### 2. データ保護型
**例**: 「パスワードを平文で保存しない」

**実装確認**:
```csharp
// ❌ 仕様違反コード（検出すべきパターン）
user.Password = inputPassword; // 平文保存

// ✅ 仕様準拠コード
user.PasswordHash = _passwordHasher.HashPassword(user, inputPassword);
```

#### 3. 処理省略型
**例**: 「削除時に物理削除は行わない（論理削除のみ）」

**テスト手法**:
```csharp
[Fact]
public async Task Delete_ShouldPerformSoftDelete_NotHardDelete()
{
    // Arrange
    var entityId = Guid.NewGuid();
    
    // Act
    await _repository.Delete(entityId);
    
    // Assert - データベースに存在し、削除フラグがtrueであることを確認
    var entity = await _dbContext.Entities
        .IgnoreQueryFilters() // 論理削除フィルタを無視
        .FirstOrDefaultAsync(e => e.Id == entityId);
    
    Assert.NotNull(entity); // 物理的に存在
    Assert.True(entity.IsDeleted); // 論理削除フラグ
}
```

## 🔍 仕様項番トレーサビリティ

### 実装と仕様の対応表管理

各Stepの組織設計ファイルに以下のセクションを追加：

```markdown
## 📋 仕様準拠マトリックス

| 機能 | 仕様書項番 | 実装ファイル | テストファイル | 準拠状態 |
|------|------------|--------------|----------------|----------|
| ログイン機能 | 2.1.1 | AuthService.cs | AuthServiceTests.cs | ✅ |
| ロックアウト非実装 | 2.1.1 | AuthService.cs | AuthServiceTests.cs#L45 | ✅ |
| パスワード変更 | 2.1.2 | UserService.cs | UserServiceTests.cs | 🚧 |
| 初回ログイン強制 | 2.1.2 | LoginHandler.cs | - | ❌ |
```

### 仕様変更影響分析

仕様変更時の影響範囲特定プロセス：

```
1. 変更された仕様書項番を特定
2. 仕様準拠マトリックスから影響を受ける実装を検索
3. 関連するテストケースをリストアップ
4. 修正計画を立案
```

## 📊 仕様準拠チェックリスト

### Step開始時チェックリスト
```
□ 該当する仕様書セクションを特定済み
□ 肯定的仕様（実装すべき機能）をリストアップ済み
□ 否定的仕様（実装してはいけない機能）をリストアップ済み
□ ビジネスルールを完全に理解済み
□ 仕様準拠マトリックスを作成済み
```

### テスト設計時チェックリスト
```
□ すべての肯定的仕様に対するテストケースを作成
□ すべての否定的仕様に対する検証テストを作成
□ ビジネスルールの境界値テストを作成
□ テストケース名に仕様書項番を含める
```

### Step終了時チェックリスト
```
□ 仕様準拠マトリックスのすべての項目が✅
□ 否定的仕様の非実装を確認済み
□ コード内に仕様書項番コメントを記載済み
□ 仕様逸脱リスクがないことを確認済み
```

## 🏢 組織設計への統合

### 仕様準拠監査役の責任
- 仕様書の深い理解と解釈
- 実装と仕様の整合性確認
- 否定的仕様の非実装確認
- 仕様準拠マトリックスの維持管理

### チーム全体での仕様理解
- Step開始時の仕様読み合わせ会
- 曖昧な仕様の解釈統一
- 仕様変更時の影響範囲共有

## 🔄 継続的改善

### 仕様準拠の定量評価
- 仕様準拠率: 実装済み機能 / 仕様書記載機能
- 否定的仕様遵守率: 非実装確認済み / 否定的仕様総数
- 仕様項番カバレッジ: 紐付け済み実装 / 全実装

### Phase完了時の振り返り
- 仕様理解の課題と改善策
- 仕様書の曖昧さと明確化提案
- 仕様準拠プロセスの改善点

---

**関連文書**:
- [組織管理運用マニュアル](./組織管理運用マニュアル.md)
- [テスト戦略ガイド](./テスト戦略ガイド.md)
- [機能仕様書](/Doc/01_Requirements/機能仕様書.md)